src/lib-imap/imap-capability-list.c

```diff
@@ -0,0 +1,153 @@
+		const char *name, enum imap_capability_visibility flags)
+{
+	/* make sure it isn't already in the list */
+	i_assert(imap_capability_list_find(capability_list, name) == NULL);
+
+	struct imap_capability capability;
+	/* initialize the capability object */
+	capability.name = p_strdup(capability_list->pool, name);
+	capability.flags = flags;
+
+	/* append new capability object to array */
+	array_append(&capability_list->imap_capabilities, &capability, 1);
+}
+
+/* remove an entry from the imap capability list */
+void imap_capability_list_remove(struct imap_capability_list *capability_list,
+		const char *name)
+{
+	/* iterate array of capabilities and find our capability */
+	unsigned int i, count;
+	const struct imap_capability *cap =
+		array_get(&capability_list->imap_capabilities, &count);
+	for (i = 0; i < count; i++) {
+		if (strcasecmp(cap[i].name, name)) {
+			array_delete(&capability_list->imap_capabilities, i, 1);
+			return;
+		}
+	}
+	i_unreached();
+}
+
+/* split a string by space and add each entry to the imap capability list */
+void imap_capability_list_append_string(struct imap_capability_list *capability_list,
+		const char *name_list)
+{
+	T_BEGIN {
+		/* split the name list by space */
+		const char **strings = t_strsplit_spaces(name_list, " ");
+		/* iterate each name and append it to list as ALWAYS */
+		while (*strings != NULL) {
+			imap_capability_list_add(capability_list, *strings, 
+					IMAP_CAP_VISIBILITY_ALWAYS);
+			strings++;
+		}
+	} T_END;
+}
+
+/* capability list bsearch comparison callback */
+static int imap_capability_bsearch(const char *name,
+		const struct imap_capability *cap)
+{
+	return strcasecmp(name, cap->name);
+}
+
+/* capability list bsearch 'find' function */
+struct imap_capability *imap_capability_list_find(struct imap_capability_list *capability_list,
+		const char *name)
+{
+	return array_bsearch(&capability_list->imap_capabilities,
+			name, imap_capability_bsearch);
+}
+
+/* append the capability list to cap_str based on given IMAP_CAP_VISIBILITY_ flags */
+void imap_capability_list_get_capability(struct imap_capability_list *capability_list, 
+		string_t *cap_str, uint32_t flags)
+{
+	/* iterate capability list and append each capability if conditions met */
+	const struct imap_capability *capp;
+	array_foreach(&capability_list->imap_capabilities, capp) {
+		/* local flags variable will get filled with all the flags
+		   that matched the current conditions -- then at the end
+		   if FLAG_REQUIRE_ALL is present the local flags variable
+		   is compared against the capability flags to check for a
+		   perfect match */
+		uint32_t n = 0;
+		if (capp->flags) {
+			/* grab the flags of the current cap into a tmp variable */
+			uint32_t tmp_flags = capp->flags;
+			/* strip off the REQUIRE_ALL and any flags that don't match */
+			tmp_flags &= (flags & ~IMAP_CAP_VISIBILITY_REQUIRE_ALL);
+
+			bool add = FALSE;
+			/*  now check if IMAP_CAP_VISIBILITY_REQUIRE_ALL is present */
```

Comment:
here and below these comments have two spaces after /* for some reason.. Also in some other places:

```
lib-imap/imap-capability-list.c:                        /*  now check if IMAP_CAP_VISIBILITY_REQUIRE_ALL is present */
lib-imap/imap-capability-list.c:                                /*  did we match all? */
lib-imap/imap-capability-list.c:                        /*  no IMAP_CAP_VISIBILITY_REQUIRE_ALL - did we match any flags? */
lib-imap/imap-capability-list.c:                        /*  should we add this entry? */
lib-imap/imap-capability-list.h:/*  the initial size of the capability list
lib-imap/imap-capability-list.h:/*  The amount to grow the capability list
lib-imap/imap-capability-list.h:/*  This structure describes a capability word that appears
```